spanner/spanner_snippets/spanner/spanner_batch_write.go

```
@@ -0,0 +1,73 @@
+)
+
+// batchWrite demonstrates writing mutations to a Spanner database through
+// BatchWrite API - https://pkg.go.dev/cloud.google.com/go/spanner#Client.BatchWrite
+func batchWrite(w io.Writer, db string) error {
+	// db := "projects/my-project/instances/my-instance/databases/my-database"
+	ctx := context.Background()
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Database is assumed to exist - https://cloud.google.com/spanner/docs/getting-started/go#create_a_database
+	singerColumns := []string{"SingerId", "FirstName", "LastName"}
+	albumColumns := []string{"SingerId", "AlbumId", "AlbumTitle"}
+	mutationGroups := make([]*spanner.MutationGroup, 2)
+
+	mutationGroup1 := []*spanner.Mutation{
+		spanner.InsertOrUpdate("Singers", singerColumns, []interface{}{16, "Scarlet", "Terry"}),
+	}
+	mutationGroups[0] = &spanner.MutationGroup{Mutations: mutationGroup1}
+
+	mutationGroup2 := []*spanner.Mutation{
+		spanner.InsertOrUpdate("Singers", singerColumns, []interface{}{17, "Marc", ""}),
+		spanner.InsertOrUpdate("Singers", singerColumns, []interface{}{18, "Catalina", "Smith"}),
+		spanner.InsertOrUpdate("Albums", albumColumns, []interface{}{17, 1, "Total Junk"}),
+		spanner.InsertOrUpdate("Albums", albumColumns, []interface{}{18, 2, "Go, Go, Go"}),
+	}
+	mutationGroups[1] = &spanner.MutationGroup{Mutations: mutationGroup2}
+	iter := client.BatchWrite(ctx, mutationGroups)
+	// See https://pkg.go.dev/cloud.google.com/go/spanner#BatchWriteResponseIterator.Do
+	doFunc := func(response *sppb.BatchWriteResponse) error {
+		if err = status.ErrorProto(response.GetStatus()); err == nil {
+			fmt.Fprintf(w, "Mutation group indexes %v have been applied with commit timestamp %v",
+				response.GetIndexes(), response.GetCommitTimestamp())
+		} else {
+			fmt.Fprintf(w, "Mutation group indexes %v could not be applied with error %v",
+				response.GetIndexes(), err)
+		}
+		// Return an actual error as needed.
+		return nil
```

We are always returning `nil` which will always succeed the `runSample` method even in case of failures in batch write sample. We should return the error instead to indicate that the test failed.
Currently it might fail the assertion but ideally the sample invocation should fail.

---

Batch write allows for partial failure and thus did it this way - process all response items and collect successes and failures. Partial failure may not mean failure for the client app.